### DISCUSSION TITLE
Clerk integration - The inferred type of ... cannot be named without a reference to...

### DISCUSSION THREAD

**Discussion by adartaud** [t3-oss/create-t3-turbo#1132] [Q&A]
Hey there,

I am replacing Next-auth with Clerk.

Here is the creation of TRPC context in the `api` package

```ts
export const createTRPCContext = (opts: { headers: Headers }) => {
  return {
    db,
    auth: auth(),
    ...opts,
  }
}
```

However, by doing so I get the following type error:
```
The inferred type of 'createTRPCContext' cannot be named without a reference to '../../../node_modules/@clerk/nextjs/dist/types/server/protect'. This is likely not portable. A type annotation is necessary.
```

I managed to solve the issue by modifying the `tsconfig` of the API package, but I guess ideally I should be able to set these properties as true:
```json
{
  "compilerOptions": {
    /** TODO: this is a fix because of some Clerk types not being inferred. */
    "declaration": false,
    "declarationMap": false,
    "emitDeclarationOnly": false
  },
}
```

Here are the dependencies of the API package:

```json
"dependencies": {
    "@clerk/nextjs": "^5.2.4",
    "@trpc/server": "11.0.0-rc.441",
    "superjson": "2.2.1",
    "zod": "^3.23.8"
  }
```

Did someone already manage to integrate Clerk without this issue ?
I'm not sure if this is an issue with my configuration, t3-turbo or Clerk package.

Thanks,

**Comment by jmarbutt** [comment]
Did you figure out the solution for clerk?

**Comment by timmhirsens** [comment]
I also had this issue when migrating an existing t3 app to this structure and all i came up with is this lousy workaround:

```typescript
type SignedInAuthObject = {
  sessionId: string;
  userId: string;
};

type SignedOutAuthObject = {
  sessionId: null;
  userId: null;
};

type AuthObject = SignedInAuthObject | SignedOutAuthObject;

export const createTRPCContext = async (opts: {
  headers: Headers;
  auth: AuthObject;
  token: string | null;
}) => {
  return {
    db,
    ...opts,
  };
};
```

Basically defining your own types for the auth property... Works, if you dont need any of the other properties.

**Comment by nichita-pasecinic** [comment]
+1

**Comment by juliusmarminge** [comment]
You should open an issue in Clerk for this. Basically they don't export enough of their types to support declaration: true.
